ibc_public/utils_connectivity.py

```diff
@@ -75,82 +213,546 @@ def get_ses_modality(task):
+
+    # select data based current task and whatever is being classified
+    # get classes and groups
+    data, task_label, classes, groups = _select_data(data, classify, task)
+    # pick specific connectome to classify based on Sconnectivity measure
+    connectomes = np.array(data[connectivity_measure].values.tolist())
+    # get train and test splits
+    splits = np.array(
+        [*cv.split(connectomes, classes, groups)], dtype="object"
+    )
+    # too many splits for Tasks from LPGO scheme, so randomly select some
+    if classify == "Tasks":
+        rng = np.random.default_rng(1)
+        select_splits = rng.choice(len(splits), cv_splits, replace=False)
+        splits = splits[select_splits]
+    # plot the train/test cross-validation splits
+    plot_cv_indices(
+        splits,
+        connectomes,
+        classes,
+        groups,
+        results_dir,
+        task_label,
+        connectivity_measure,
+    )
+    # initialize lists to store results
+    results = {
+        "accuracy": [],
+        "auc": [],
+        "predicted_class": [],
+        "true_class": [],
+        "connectivity": [],
+        "classes": [],
+        "groups": [],
+        "train_sets": [],
+        "test_sets": [],
+        "task_label": [],
+    }
+    results = cross_validate(
+        connectomes,
+        classes,
+        splits,
+        task_label,
+        connectivity_measure,
+        results,
+        classify,
+    )
+
+    results_df = pd.DataFrame(results)
+    # save the results
+    results_file = os.path.join(
+        results_dir,
+        f"{task_label}_{connectivity_measure}_results.csv",
+    )
+    results_df.to_csv(results_file)
+    # do_plots(results, results_dir, classify)
+
+    return results_df
+
+
+def do_plots(results, results_dir, classify):
+    results = chance_level(results)
+    print("Plotting results...")
+    # plot accuracy as barplots
+    plot_accuracy(results, results_dir, classify)
+    # plot confusion matrices
+    plot_confusion(results, results_dir)
+
+
+def chance_level(df):
+    dfs = []
+    tasks = df["task_label"].unique()
+    conn_measures = df["connectivity"].unique()
+    for task in tasks:
+        for conn_measure in conn_measures:
+            task_df = df[
+                (df["connectivity"] == conn_measure)
+                & (df["task_label"] == task)
+            ]
+            classes = []
+            n_splits = len(task_df)
+            for _, row in task_df.iterrows():
+                classes.extend(task_df["true_class"].tolist())
+                classes.extend(task_df["predicted_class"].tolist())
+            classes = np.concatenate(classes)
+            classes = np.asarray(classes)
+            classes = np.unique(classes)
+            task_df["labels"] = [classes for split in range(n_splits)]
+            task_df["n_labels"] = [len(classes) for split in range(n_splits)]
+            task_df["chance"] = [1 / len(classes) for split in range(n_splits)]
```

Comment:
you assume that the classes are balanced... which may not be the case.

Reply:
Maybe better to rely on a DummyClassifier

Reply:
Ok yes! That makes sense, I'll do that